tools/storage/segment.py

```diff
@@ -0,0 +1,318 @@
+        headers = {}
+        for _ in range(0, header_count):
+            hkey = None
+            hvalue = None
+            hkey_length, v = read_varlong(b, v)
+            if hkey_length > 0:
+                hkey = b[v:v + hkey_length]
+                v += hkey_length
+            hvalue_length, v = read_varlong(b, v)
+            if hvalue_length > 0:
+                hvalue = b[v:v + hvalue_length]
+                v += hvalue_length
+            if hkey is not None:
+                headers[hkey] = hvalue
+        return Record(size, record_attr, timestamp_delta, offset_delta, key,
+                      headers, value)
+
+    @staticmethod
+    def records_from_bytes(bytes, count):
+        records = []
+        for _ in range(0, count):
+            record = Record.from_bytes(bytes)
+            size = record.size + 1
+            bytes = bytes[size:]
+            records.append(record)
+        return records
+
+
+class Batch:
+    """ Batch of Records in a Segment file """
+
+    def __init__(self, index, header, records):
+        self.index = index
+        self.header = header
+        self.records = records
+        self.__verify_crc()
+
+    def __verify_crc(self):
+        header_crc_bytes = struct.pack(
+            "<" + HDR_FMT_RP_PREFIX_NO_CRC + HDR_FMT_CRC, *self.header[1:])
+        header_crc = crc32c.crc32c(header_crc_bytes)
+        if self.header.header_crc != header_crc:
+            raise CorruptBatchError(self)
+        crc = crc32c.crc32c(self._crc_header_be_bytes())
+        crc = crc32c.crc32c(self.records, crc)
+        if self.header.crc != crc:
+            raise CorruptBatchError(self)
+
+    def last_offset(self):
+        return self.header.base_offset + self.header.record_count - 1
+
+    def _crc_header_be_bytes(self):
+        # encode header back to big-endian for crc calculation
+        return struct.pack(">" + HDR_FMT_CRC, *self.header[5:])
+
+    def parse_records(self):
+        bytes = self.records
+        if self.header.type != 1:
+            return []
+        try:
+            if self.header.attrs == 0:
+                pass  # uncompressed
+            elif self.header.attrs == 1:
+                logger.debug("Decompressing gzip batch")
+                bytes = gzip.decompress(self.records)
+            elif self.header.attrs == 2:
+                raise NotImplementedError(
+                    "Unsupported compression type: Snappy")
+            elif self.header.attrs == 3:
+                logger.debug("Decompressing LZ4 batch")
+                bytes = lz4.frame.decompress(self.records)
+            elif self.header.attrs == 4:
+                logger.debug("Decompressing Zstandard batch")
+                bytes = zstd.decompress(self.records)
+            else:
+                raise NotImplementedError(
+                    f"Unknown compression type: {self.header.attrs}")
+            return Record.records_from_bytes(bytes, self.header.record_count)
```

Comment:
@jrkinley you might consider making the decoding of the records which are the only user of variable length encoding optional. that will speed things up, and I would assume that decoding the user payloads is going to be less common.

fyi: there are some tricks in https://github.com/dpkp/kafka-python/blob/f0a57a6a20a3049dc43fbf7ad9eab9635bd2c0b0/kafka/record/default_records.py for performance related to working on a raw byte buffers.